[wip] Switch to C++14

smessmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#27864 [wip] Switch to C++14**
* #30359 [wip] update to xcode 9.4, cuda 10.0, cudnn 7.6.5

Differential Revision: [D17908478](https://our.internmc.facebook.com/intern/diff/D17908478/)

smessmer committed:
[wip] Switch to C++14



Differential Revision: [D17908478](https://our.internmc.facebook.com/intern/diff/D17908478/)

[ghstack-poisoned]

smessmer committed:
Update on "[wip] Switch to C++14"


Differential Revision: [D17908478](https://our.internmc.facebook.com/intern/diff/D17908478/)

[ghstack-poisoned]

smessmer committed:
Update on "[wip] Switch to C++14"


Differential Revision: [D17908478](https://our.internmc.facebook.com/intern/diff/D17908478/)

[ghstack-poisoned]

soumith reviewed (approved):
conditional on CI passing.

@peterjc123 anything to know about C++14 and Windows incompatibility, now is the time to speak :)

smessmer committed:
Update on "[wip] Switch to C++14"


Differential Revision: [D17908478](https://our.internmc.facebook.com/intern/diff/D17908478/)

[ghstack-poisoned]

peterjc123 commented:
@soumith According to this [table](https://docs.microsoft.com/en-us/cpp/overview/visual-cpp-language-conformance?view=vs-2019), all the C++ 14 features are available in VS 2017. `/std:c++14` is automatically enabled in MSVC 2017, so there is actually nothing we should do to add C++ 14 support. More details could be found [here](https://docs.microsoft.com/en-us/cpp/build/reference/std-specify-language-standard-version?view=vs-2019).

bddppq commented on cmake/Dependencies.cmake:
```diff
@@ -968,6 +968,7 @@ if(USE_ROCM)
     list(APPEND HIP_CXX_FLAGS -Wno-duplicate-decl-specifier)
     list(APPEND HIP_CXX_FLAGS -DCAFFE2_USE_MIOPEN)
     list(APPEND HIP_CXX_FLAGS -DTHRUST_DEVICE_SYSTEM=THRUST_DEVICE_SYSTEM_HIP)
+    list(APPEND HIP_CXX_FLAGS -std=gnu++14)
```
`-std=gnu++14` -> `-std=c++14`, hcc doesn't support gnu extensions.

smessmer committed:
Update on "[wip] Switch to C++14"


Differential Revision: [D17908478](https://our.internmc.facebook.com/intern/diff/D17908478/)

[ghstack-poisoned]

smessmer committed:
Update on "[wip] Switch to C++14"


Differential Revision: [D17908478](https://our.internmc.facebook.com/intern/diff/D17908478/)

[ghstack-poisoned]

smessmer committed:
Update on "[wip] Switch to C++14"


Differential Revision: [D17908478](https://our.internmc.facebook.com/intern/diff/D17908478/)

[ghstack-poisoned]

smessmer committed:
Update on "[wip] Switch to C++14"


Differential Revision: [D17908478](https://our.internmc.facebook.com/intern/diff/D17908478/)

[ghstack-poisoned]

smessmer committed:
Update on "[wip] Switch to C++14"


Differential Revision: [D17908478](https://our.internmc.facebook.com/intern/diff/D17908478/)

[ghstack-poisoned]

smessmer committed:
Update on "[wip] Switch to C++14"


Differential Revision: [D17908478](https://our.internmc.facebook.com/intern/diff/D17908478/)

[ghstack-poisoned]

smessmer committed:
Update on "[wip] Switch to C++14"


Differential Revision: [D17908478](https://our.internmc.facebook.com/intern/diff/D17908478/)

[ghstack-poisoned]

smessmer committed:
Update on "[wip] Switch to C++14"


Differential Revision: [D17908478](https://our.internmc.facebook.com/intern/diff/D17908478/)

[ghstack-poisoned]

smessmer committed:
Update on "[wip] Switch to C++14"


Differential Revision: [D17908478](https://our.internmc.facebook.com/intern/diff/D17908478/)

[ghstack-poisoned]

smessmer committed:
Update on "[wip] Switch to C++14"


Differential Revision: [D17908478](https://our.internmc.facebook.com/intern/diff/D17908478/)

[ghstack-poisoned]

smessmer committed:
Update on "[wip] Switch to C++14"


Differential Revision: [D17908478](https://our.internmc.facebook.com/intern/diff/D17908478/)

[ghstack-poisoned]

smessmer committed:
Update on "[wip] Switch to C++14"


Differential Revision: [D17908478](https://our.internmc.facebook.com/intern/diff/D17908478/)

[ghstack-poisoned]

smessmer commented:
This is not merged yet, there was an error in the export. Re-created this PR in  https://github.com/pytorch/pytorch/pull/30406.
